File: kafka-test-scala/src/main/scala/Consumer.scala
package com.example.kafka

import org.apache.kafka.clients.consumer.{ConsumerConfig, KafkaConsumer}
import org.apache.kafka.common.errors.WakeupException
import org.apache.kafka.common.serialization.StringDeserializer
import org.slf4j.LoggerFactory
import scala.jdk.CollectionConverters._
import java.time.Duration
import java.util.Properties
import java.util.Collections
import java.util.concurrent.atomic.AtomicBoolean

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.scala.DefaultScalaModule

object Consumer {
  private val logger = LoggerFactory.getLogger(getClass)
  private val objectMapper = new ObjectMapper().registerModule(DefaultScalaModule)

  def main(args: Array[String]): Unit = {
    val bootstrapServers = sys.env.getOrElse("KAFKA_BOOTSTRAP_SERVERS", "127.0.0.1:9094")
    val topic = sys.env.getOrElse("TOPIC_NAME", "my-topic-1")
    val groupId = sys.env.getOrElse("CONSUMER_GROUP_ID", "my-group-1")
    val offsetReset = sys.env.getOrElse("AUTO_OFFSET_RESET", "latest")

    logger.info(s"🛠️ Connecting KafkaConsumer to topic '$topic' at '$bootstrapServers' (group: '$groupId')...")

    val props = new Properties()
    props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers)
    props.put(ConsumerConfig.GROUP_ID_CONFIG, groupId)
    props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, classOf[StringDeserializer].getName)
    props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, classOf[StringDeserializer].getName)
    props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, offsetReset)
    props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "true")
    props.put(ConsumerConfig.SESSION_TIMEOUT_MS_CONFIG, "6000") // Fail fast

    // Force IPv4 for local dev
    System.setProperty("java.net.preferIPv4Stack", "true")

    val consumer = new KafkaConsumer[String, String](props)
    val closed = new AtomicBoolean(false)

    Runtime.getRuntime.addShutdownHook(new Thread(() => {
      logger.info("🛑 Shutting down gracefully...")
      closed.set(true)
      consumer.wakeup()
    }))

    try {
      consumer.subscribe(Collections.singletonList(topic))
      logger.info("✅ KafkaConsumer connected. Waiting for new messages...")
      logger.info("📥 Listening for messages...")

      while (!closed.get()) {
        val records = consumer.poll(Duration.ofMillis(100))
        for (record <- records.asScala) {
          try {
            val nowMillis = System.currentTimeMillis()
            val nowSeconds = nowMillis / 1000.0

            val json = objectMapper.readTree(record.value())
            val messageId = json.get("message_id").asLong()
            val eventTime = json.get("event_time").asDouble()
            val content = json.get("content").asText()

            val latency = nowSeconds - eventTime

            logger.info(f"📨 New message: [ID=$messageId] $content at $eventTime%.3f")
            logger.info(f"⏱️ Latency: $latency%.3f seconds")
          } catch {
            case e: Exception => logger.warn(s"⚠️ Error parsing message: ${e.getMessage} - payload: ${record.value()}")
          }
        }
      }
    } catch {
      case _: WakeupException => if (!closed.get()) throw new RuntimeException("Unexpected wakeup")
      case e: Exception => logger.error("❌ KafkaConsumer error: ", e)
    } finally {
      consumer.close()
      logger.info("✅ KafkaConsumer closed.")
    }
  }
}

File: kafka-test-java/src/main/java/com/example/kafka/Consumer.java
package com.example.kafka;

import java.io.File;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.Duration;
import java.util.Collections;
import java.util.Properties;
import java.util.concurrent.atomic.AtomicBoolean;
import org.apache.avro.Schema;
import org.apache.avro.generic.GenericDatumReader;
import org.apache.avro.generic.GenericRecord;
import org.apache.avro.io.BinaryDecoder;
import org.apache.avro.io.DecoderFactory;
import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.clients.consumer.ConsumerRecords;
import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.kafka.common.errors.WakeupException;
import org.apache.kafka.common.serialization.ByteArrayDeserializer;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class Consumer {

    private static final Logger logger = LoggerFactory.getLogger(
        Consumer.class
    );

    private static Schema loadSchema() throws Exception {
        Path[] paths = {
            Paths.get("../schemas/message.avsc"),
            Paths.get("/app/schemas/message.avsc"),
            Paths.get("schemas/message.avsc"),
        };

        for (Path p : paths) {
            if (Files.exists(p)) {
                return new Schema.Parser().parse(p.toFile());
            }
        }
        throw new RuntimeException("Schema file not found in paths");
    }

    public static void main(String[] args) {
        try {
            Schema schema = loadSchema();
            logger.info("Loaded Avro schema: {}", schema.getFullName());

            String bootstrapServers = System.getenv().getOrDefault(
                "KAFKA_BOOTSTRAP_SERVERS",
                "127.0.0.1:9094"
            );
            String topic = System.getenv().getOrDefault(
                "TOPIC_NAME",
                "my-topic-1"
            );
            String groupId = System.getenv().getOrDefault(
                "CONSUMER_GROUP_ID",
                "my-group-1"
            );
            String offsetReset = System.getenv().getOrDefault(
                "AUTO_OFFSET_RESET",
                "latest"
            );

            logger.info(
                "🛠️ Connecting KafkaConsumer to topic '{}' at '{}' (group: '{}')...",
                topic,
                bootstrapServers,
                groupId
            );

            Properties props = new Properties();
            props.put(
                ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG,
                bootstrapServers
            );
            props.put(ConsumerConfig.GROUP_ID_CONFIG, groupId);
            props.put(
                ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG,
                StringDeserializer.class.getName()
            );
            props.put(
                ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,
                ByteArrayDeserializer.class.getName()
            );
            props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, offsetReset);
            props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "true");
            props.put(ConsumerConfig.SESSION_TIMEOUT_MS_CONFIG, "6000"); // Fail fast if consumer dies

            // Use IPv4 if possible to avoid connection issues on localhost
            System.setProperty("java.net.preferIPv4Stack", "true");

            KafkaConsumer<String, byte[]> consumer = new KafkaConsumer<>(props);

            // Graceful shutdown
            final AtomicBoolean closed = new AtomicBoolean(false);
            Runtime.getRuntime().addShutdownHook(
                new Thread(() -> {
                    logger.info("🛑 Shutting down gracefully...");
                    closed.set(true);
                    consumer.wakeup();
                })
            );

            GenericDatumReader<GenericRecord> reader = new GenericDatumReader<>(
                schema
            );

            try {
                consumer.subscribe(Collections.singletonList(topic));
                logger.info(
                    "✅ KafkaConsumer connected. Waiting for new messages..."
                );
                logger.info("📥 Listening for messages...");

                while (!closed.get()) {
                    ConsumerRecords<String, byte[]> records = consumer.poll(
                        Duration.ofMillis(100)
                    );
                    for (ConsumerRecord<String, byte[]> record : records) {
                        try {
                            BinaryDecoder decoder =
                                DecoderFactory.get().binaryDecoder(
                                    record.value(),
                                    null
                                );
                            GenericRecord avroRecord = reader.read(
                                null,
                                decoder
                            );

                            long messageId = (long) avroRecord.get(
                                "message_id"
                            );
                            double eventTime = (double) avroRecord.get(
                                "event_time"
                            );
                            String content = avroRecord
                                .get("content")
                                .toString();

                            long nowMillis = System.currentTimeMillis();
                            double nowSeconds = nowMillis / 1000.0;
                            double latency = nowSeconds - eventTime;

                            logger.info(
                                "📨 New message: [ID={}] {} at {}",
                                messageId,
                                content,
                                String.format("%.3f", eventTime)
                            );
                            logger.info(
                                "⏱️ Latency: {} seconds",
                                String.format("%.3f", latency)
                            );
                        } catch (Exception e) {
                            logger.warn(
                                "⚠️ Error parsing message: {} - offset: {}",
                                e.getMessage(),
                                record.offset()
                            );
                        }
                    }
                }
            } catch (WakeupException e) {
                if (!closed.get()) throw e;
            } catch (Exception e) {
                logger.error("❌ KafkaConsumer error: ", e);
            } finally {
                consumer.close();
                logger.info("✅ KafkaConsumer closed.");
            }
        } catch (Exception e) {
            logger.error("❌ Fatal error: ", e);
        }
    }
}
